link/netkit_test.go

```diff
@@ -0,0 +1,103 @@
+package link
+
+import (
+	"fmt"
+	"net"
+	"testing"
+
+	"github.com/cilium/ebpf"
+	"github.com/cilium/ebpf/internal/testutils"
+	"github.com/go-quicktest/qt"
+	"github.com/jsimonetti/rtnetlink/v2"
+	"github.com/jsimonetti/rtnetlink/v2/driver"
+	"golang.org/x/sys/unix"
+)
+
+func TestAttachNetkit(t *testing.T) {
+	testutils.SkipOnOldKernel(t, "6.7", "Netkit Device")
+
+	prog := mustLoadProgram(t, ebpf.SchedCLS, ebpf.AttachNetkitPrimary, "")
+	link, _ := mustAttachNetkit(t, prog, ebpf.AttachNetkitPrimary)
+
+	testLink(t, link, prog)
+}
+
+func TestNetkitAnchor(t *testing.T) {
+	testutils.SkipOnOldKernel(t, "6.7", "Netkit Device")
+
+	a := mustLoadProgram(t, ebpf.SchedCLS, ebpf.AttachNetkitPrimary, "")
+	b := mustLoadProgram(t, ebpf.SchedCLS, ebpf.AttachNetkitPrimary, "")
+
+	linkA, iface := mustAttachNetkit(t, a, ebpf.AttachNetkitPrimary)
+
+	programInfo, err := a.Info()
+	qt.Assert(t, qt.IsNil(err))
+	programID, _ := programInfo.ID()
+
+	linkInfo, err := linkA.Info()
+	qt.Assert(t, qt.IsNil(err))
+	linkID := linkInfo.ID
+
+	for _, anchor := range []Anchor{
+		Head(),
+		Tail(),
+		BeforeProgram(a),
+		BeforeProgramByID(programID),
+		AfterLink(linkA),
+		AfterLinkByID(linkID),
+	} {
+		t.Run(fmt.Sprintf("%T", anchor), func(t *testing.T) {
+			linkB, err := AttachNetkit(NetkitOptions{
+				Program:   b,
+				Attach:    ebpf.AttachNetkitPrimary,
+				Interface: iface,
+				Anchor:    anchor,
+			})
+			qt.Assert(t, qt.IsNil(err))
+			qt.Assert(t, qt.IsNil(linkB.Close()))
+		})
+	}
+}
+
+func mustAttachNetkit(tb testing.TB, prog *ebpf.Program, attachType ebpf.AttachType) (Link, int) {
+	var err error
+	conn, err := rtnetlink.Dial(nil)
+	qt.Assert(tb, qt.IsNil(err))
+
+	layer2 := driver.NetkitModeL2
+	blackhole := driver.NetkitPolicyDrop
+	err = conn.Link.New(&rtnetlink.LinkMessage{
+		Family: unix.AF_UNSPEC,
+		Index:  1000,
```

Comment:
This means we can't call the function more than once per test or concurrently.

Is there a way to get the kernel to allocate an ID + retrieve the ID from the netlink answer somehow?

Reply:
If we do not provide the index the kernel will assign the index, the same goes for the name, just like the ip command line tool without giving any arguments. We are getting the link with `net.InterfaceByName` so we do not need to provide the index. But even if we get rid of the index we still have a hard-coded interface name. I am not sure what we can do about that. If we do not give the name the kernel will assign names starting from nk0 and nk1, and whatever available next.

Reply:
Could we get the interface index from the calling test function? It would be the caller's responsibility to give a unique index for every call. Also, we would not need to call `net.InterfaceByName` nor return the index from `net.InterfaceByName`. Also, conn can come from the caller so that if we need to call the function more than once we would not need to create new connections every time. What do you think?

Reply:
> Could we get the interface index from the calling test function? It would be the caller's responsibility to give a unique index for every call.

I like this, seems straightforward enough.

>  Also, conn can come from the caller so that if we need to call the function more than once we would not need to create new connections every time. What do you think?

That would make it more cumbersome to call (because we need another func to dial the conn). I'd just dial a new conn as needed until it turns out that doesn't work for some reason.

Reply:
Why not generate a random string at the start of the function and use it as the interface name?

Reply:
Can do, ultimately this isn't going to happen often I'd guess. So manually allocating indices is probably not too much of a hassle.

Reply:
Push changes.

Reply:
Ended up adding a small sequence generator which allocates ids automatically.